### PR TITLE
WebHost/Core/Lingo: Render option documentation as reStructuredText in the WebView

### PR DESCRIPTION
**Pull request by nex3**
## What is this fixing or adding?

This allows Options to use the standard Python documentation format,
while producing much nicer-looking documentation in the WebView with
things like emphasis, lists, and so on.

While this behavior is on by default for new worlds, this also (at
@NewSoupVi's suggestion) adds a list of worlds that are opted into the
old "plain text" behavior. World authors can remove their worlds from
this list and update their Option documentation at their leisure.

### HTML Tooltips

This also updates the tooltip infrastructure to support full HTML
tooltips, structured like so:

```html
<div class="tooltip-container">
  (?)
  <div class="tooltip">Tooltip contents</div>
</div>
```

Plain text tooltips continue to work using the `data-tooltip`
attribute.

## How was this tested?

I updated the Lingo documentation to use the new behavior, and
verified that other worlds' output doesn't change.

## If this makes graphical changes, please attach screenshots.

**Old look:**

image

**New look:**

image

### CODE REVIEW

File: worlds/legacy_world_features.py
Comment (black-sliver) [inline worlds/legacy_world_features.py:14]: This breaks APWorlds. Please don't do this. Let the world opt in.

File: WebHostLib/options.py
Comment (black-sliver) [inline WebHostLib/options.py:81]: We probably want to avoid raw html for several reasons
```python
text = """.. raw:: html

    <script>alert('pwned')</script>"""
publish_parts(text, writer_name="html", settings=None, settings_overrides={"output_encoding": "unicode"})["body"]
```
returns "<script>alert('pwned')</script>"

Security may be a concern, since the option export could be "safe" (i.e. jsonworlds), but even if it wasn't, we don't want to have people put raw html there to work around issues instead of fixing them in core.

File: Options.py
Comment (ThePhar) [inline Options.py:129]: Rather than attach this to the `Option` class itself, wouldn't it make more sense to make this an attribute on `WebWorld` itself that applies to all options?

I don't see a reason why anyone migrating to reST formatting wouldn't just update all their doccomments, and depending on which becomes default (plain text should be default imo), if a dev decides to use the other, they'll have to set it on every option class.

Comment (ThePhar) [inline Options.py:129]: Rather than attach this to the `Option` class itself, wouldn't it make more sense to make this an attribute on `WebWorld` itself that applies to all options?

I don't see a reason why anyone migrating to reST formatting wouldn't just update all their doccomments, and depending on which becomes default (plain text should be default imo), if a dev decides to use the other, they'll have to set it on every option class.

File: worlds/lingo/options.py
Comment (ThePhar) [inline worlds/lingo/options.py:1]: This should probably have review from @hatkirby since it's touching their world.

File: Options.py
Comment (ThePhar) [inline Options.py:1155]: This is not at all accurate to how progression balancing works, so I disagree with this doc comment change.

File: worlds/lingo/options.py
Comment (hatkirby) [inline worlds/lingo/options.py:123]: typo

**Review by hatkirby — CHANGES_REQUESTED** [review]
Overall looks great from the Lingo side! I found two minor typos but otherwise I think this is a good change.

File: Options.py
Comment (NewSoupVi) [reply Options.py:129]: This is such a better suggestion than what I've been suggesting holy damn

**Comment by NewSoupVi** [comment]
I think I am now more on board with Phar's suggestion of making this an attribute of WebWorld, and at that point, it's so easy to opt into that I'm in agreement with Black Sliver that it should be opt-in.

File: worlds/legacy_world_features.py
Comment (NewSoupVi) [reply worlds/legacy_world_features.py:14]: Imo breaking something purely visual for an apworld during AP's 0.# days, especially when apworld don't usually worry about being displayed on WebHost (cause how would they), needs to be okay

Comment (NewSoupVi) [reply worlds/legacy_world_features.py:14]: However, with Phar's alternate suggestion of how to implement this, I'm now on board with this being opt in. Just saying right now that I disagree with your philisophy in general and it will probably come up again in the future

**Comment by ThePhar** [comment]
Oh, and if it wasn't mentioned elsewhere already, whatever the final API is proposed to look like, should also be documented in the world api documentation as well.

File: worlds/legacy_world_features.py
Comment (ThePhar) [reply worlds/legacy_world_features.py:14]: Considering the long-term goal is decoupling worlds from core, I'm against adding anything that adds special privileges to worlds currently in the core repo (it also can't be edited easily from outside the apworld, if one of these world authors releases an updated apworld that would get placed in a webhost).

So regardless of what is decided (breaking existing APworlds or not), I am firmly against this file existing.

Comment (nex3) [reply worlds/legacy_world_features.py]: To give some background here, this was intended as a combined solution for two problems:

1. This change shouldn't break the rendering for existing core worlds.
2. It's infeasible to update existing core worlds in-place because doing so would require a review from every world maintainer.

APWorlds could still easily keep the old behavior by adding `plain_text_doc = True` to their options (or `plain_text_options_doc = True` to their `World`). The idea isn't so much to give special privileges to core worlds as to work around how difficult it would be to treat them the same as APWorlds.

All that said, it sounds like the consensus is that the core maintainers prefer this to be opt-out. I worry that's not the right call in the long-term when plain text option docs provide no benefits other than backwards-compatibility, but I'll bow to your decision and make the change.

File: WebHostLib/options.py
Comment (nex3) [reply WebHostLib/options.py]: Fortunately this is an easy fix :slightly_smiling_face:

File: docs/options api.md
Comment (beauxq) [inline docs/options api.md:108]: It looks like this was not updated with the `WebWorld` change.

**Comment by LegendaryLinux** [comment]
I'm also on board with the opt-in method via WebWorld. I'll give the code some closer scrutiny tonight.

File: docs/options api.md
Comment (LegendaryLinux) [inline docs/options api.md:106]: Is this intended to be `ExampleWorld(WebWorld)`?

**Comment by LegendaryLinux** [comment]
I went ahead and looked at the code and it looks good. I'll double check the styles later tonight since I'm not at my desktop computer right now.

**Comment by LegendaryLinux** [comment]
Okay, this is approved pending the update to the documentation.

File: docs/options api.md
Comment (beauxq) [inline docs/options api.md:106]:
```suggestion
from worlds.AutoWorld import WebWorld


class ExampleWebWorld(WebWorld):
```

File: Options.py
Comment (beauxq) [inline Options.py:129]:
```suggestion
    rich_text_doc: typing.Optional[bool] = None
```

**Review by LegendaryLinux — APPROVED** [review]

**Pull request merged**
